src/mmds/src/token.rs

```diff
@@ -0,0 +1,562 @@
+    }
+
+    /// Create a new AES-GCM cipher entity.
+    fn create_cipher() -> Result<Aes256Gcm, Error> {
+        // Randomly generate a 256-bit key to be used for encryption/decryption purposes.
+        let mut key = [0u8; KEY_LEN];
+        read_into_buf_from_entropy_pool(&mut key).map_err(Error::EntropyPool)?;
+
+        // Create cipher entity to handle encryption/decryption.
+        Ok(Aes256Gcm::new(Key::from_slice(&key)))
+    }
+
+    /// Make sure to reinitialize the cipher under a new key before reaching
+    /// a count of 2^32 encrypted tokens under the same cipher entity.
+    fn check_encryption_count(&mut self) -> Result<(), Error> {
+        // Make sure no more than 2^32 - 1 tokens are encrypted under
+        // the same encryption key.
+        if self.num_encrypted_tokens == u32::MAX {
+            // Reinitialize the cipher entity under a new key.
+            self.cipher = TokenAuthority::create_cipher()?;
+            // Reset encrypted tokens count.
+            self.num_encrypted_tokens = 0;
+            warn!(
+                "MMDS's token authority entity has been reseeded. \
+                All previously created tokens are now invalid."
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        MIN_TOKEN_TTL_SECONDS <= ttl_seconds && ttl_seconds <= MAX_TOKEN_TTL_SECONDS
+    }
+
+    /// Compute expiry time in seconds by adding the time to live provided
+    /// to the current time measured in seconds.
+    fn compute_expiry(ttl_seconds: u32) -> u64 {
+        // Get time elapsed since UNIX_EPOCH.
+        let now_as_duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        // Convert elapsed time into seconds.
+        let ttl_duration = Duration::from_secs(u64::from(ttl_seconds));
+
+        // Compute expiry by adding ttl value to current time in seconds.
+        // This addition is safe because ttl is verified beforehand and
+        // can never be more than 6h.
+        let expiry_duration = now_as_duration.add(ttl_duration);
+
+        expiry_duration.as_secs()
+    }
+}
+
+/// Structure for token information.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+struct Token {
+    version: u16,
+    // Nonce or Initialization Vector.
+    iv: [u8; IV_LEN],
+    // Encrypted expire time.
+    payload: [u8; PAYLOAD_LEN],
+    // Tag returned after encryption.
+    tag: [u8; TAG_LEN],
+}
+
+impl Token {
+    /// Create a new token struct.
+    fn new(iv: [u8; IV_LEN], payload: [u8; PAYLOAD_LEN], tag: [u8; TAG_LEN]) -> Self {
+        Token {
+            version: 1,
+            iv,
+            payload,
+            tag,
+        }
+    }
+
+    /// Encode token structure into a string using base64 encoding.
+    fn base64_encode_token(&self) -> Result<String, Error> {
```

Comment:
I'd rename these functions to remove the `_token` part, as it's embedded in the struct name already